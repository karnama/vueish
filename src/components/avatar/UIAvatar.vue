<template>
    <div class="ui-avatar rounded-full text-lg overflow-hidden h-8 w-8
                text-white bg-red-500 flex flex-col justify-center items-center">
        <img v-show="src && loaded"
             :src="src"
             :alt="alt"
             @load="loaded = true"
             @error="loaded = false">
        <slot v-if="!loaded">
            <template v-if="alt">
                {{ alt.charAt(0).toUpperCase() }}
            </template>
            <template v-else>
                <span v-html="personIcon" />
            </template>
        </slot>
    </div>
</template>

<script lang="ts">
import { computed, defineComponent, ref } from 'vue';
import { getIcon } from '@/helpers';

export default defineComponent({
    name: 'UIAvatar',

    props: {
        /**
         * The alt text of the image.
         */
        alt: {
            type: String
        },

        /**
         * The source of the image.
         */
        src: {
            type: String
        }
    },

    setup(props) {
        const loaded = ref(false);
        const personIcon = getIcon('person');
        const showImg = computed(() => {
            return loaded.value && props.src;
        });

        return {
            loaded,
            personIcon,
            showImg
        };
    }
});
</script>
