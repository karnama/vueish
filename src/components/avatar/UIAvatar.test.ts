import { mount } from '@vue/test-utils';
import UIAvatar from '@components/avatar/UIAvatar.vue';

describe('UIAvatar', () => {
    it('should correctly display', () => {
        const wrapper = mount(UIAvatar);
        expect(wrapper.element).toMatchSnapshot();
    });

    it('should display the given default content if no src given', () => {
        const wrapper = mount(UIAvatar, {
            slots: {
                default: 'my-slot'
            }
        });

        expect(wrapper.text()).toBe('my-slot');
    });

    it('should display alt\'s first letter if given and source cannot load and slot not given', () => {
        const wrapper = mount(UIAvatar, {
            props: {
                alt: 'my alt text'
            }
        });

        expect(wrapper.text()).toBe('M');
    });

    it('should fall back to displaying the person icon', () => {
        const wrapper = mount(UIAvatar);

        expect(wrapper.get('svg').isVisible()).toBe(true);
    });

    it('should apply to correct rounded styles', async () => {
        const wrapper = mount(UIAvatar);

        expect(wrapper.html()).toContain('rounded-full');
        expect(wrapper.html()).not.toContain('rounded ');
        await wrapper.setProps({ squared: true });
        expect(wrapper.html()).not.toContain('rounded-full');
        expect(wrapper.html()).toContain('rounded');
    });

    it('should react to changes of the src prop', async () => {
        const wrapper = mount(UIAvatar);

        expect(wrapper.find('img').isVisible()).toBe(false);
        await wrapper.setProps({
            src: 'data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw=='
        });
        expect(wrapper.find('img').isVisible()).toBe(true);
    });
});
