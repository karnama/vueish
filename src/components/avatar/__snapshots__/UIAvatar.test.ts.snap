// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UIAvatar should correctly display 1`] = `
<div
  class="ui-avatar text-lg overflow-hidden h-8 w-8
                text-white bg-brand-400 flex flex-col justify-center items-center rounded-full"
>
  
  <span>
    <svg
      aria-label="person icon"
      fill="currentColor"
      height="18px"
      viewBox="0 0 24 24"
      width="18px"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M0 0h24v24H0V0z"
        fill="none"
      />
      <path
        d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v1c0 .55.45 1 1 1h14c.55 0 1-.45 1-1v-1c0-2.66-5.33-4-8-4z"
      />
    </svg>
  </span>
  
</div>
`;
